discovery-provider/src/tasks/index_user_bank.py

```diff
@@ -156,81 +161,104 @@ def process_transfer_instruction(
+    for user_id_account in user_id_accounts:
+        if user_id_account[1] == sender_account:
+            sender_user_id = user_id_account[0]
+        elif user_id_account[1] == receiver_account:
+            receiver_user_id = user_id_account[0]
+    if sender_user_id is None:
+        logger.error(
+            f"index_user_bank.py | ERROR: Unexpected user ids in results: {user_id_accounts}"
         )
-        pre_receiver_balance_dict = next(
-            (
-                balance
-                for balance in meta["preTokenBalances"]
-                if balance["accountIndex"] == receiver_index
-            ),
-            None,
+        return
+
+    # Find the right pre/post balances using the account indexes
+    # for the sender/receiver accounts since they aren't necessarily given in order
+    pre_sender_balance_dict = next(
+        (
+            balance
+            for balance in meta["preTokenBalances"]
+            if balance["accountIndex"] == sender_index
+        ),
+        None,
+    )
+    pre_receiver_balance_dict = next(
+        (
+            balance
+            for balance in meta["preTokenBalances"]
+            if balance["accountIndex"] == receiver_index
+        ),
+        None,
+    )
+    post_sender_balance_dict = next(
+        (
+            balance
+            for balance in meta["postTokenBalances"]
+            if balance["accountIndex"] == sender_index
+        ),
+        None,
+    )
+    post_receiver_balance_dict = next(
+        (
+            balance
+            for balance in meta["postTokenBalances"]
+            if balance["accountIndex"] == receiver_index
+        ),
+        None,
+    )
+
+    if (
+        pre_sender_balance_dict is None
+        or pre_receiver_balance_dict is None
+        or post_sender_balance_dict is None
+        or post_receiver_balance_dict is None
+    ):
+        logger.error("index_user_bank.py | ERROR: Sender or Receiver balance missing!")
+        return
+    pre_sender_balance = int(pre_sender_balance_dict["uiTokenAmount"]["amount"])
+    post_sender_balance = int(post_sender_balance_dict["uiTokenAmount"]["amount"])
+    pre_receiver_balance = int(pre_receiver_balance_dict["uiTokenAmount"]["amount"])
+    post_receiver_balance = int(post_receiver_balance_dict["uiTokenAmount"]["amount"])
+    sent_amount = pre_sender_balance - post_sender_balance
+    received_amount = post_receiver_balance - pre_receiver_balance
+    if sent_amount != received_amount:
+        logger.error(
+            f"index_user_bank.py | ERROR: Sent and received amounts don't match. Sent = {sent_amount}, Received = {received_amount}"
         )
-        post_sender_balance_dict = next(
-            (
-                balance
-                for balance in meta["postTokenBalances"]
-                if balance["accountIndex"] == sender_index
-            ),
-            None,
+        return
+
+    # If there was only 1 user bank, index as a send external transfer
```

Comment:
Might be worth noting in the comment that this has to be a send, as receives won't use the program?